Feature/protocol violation

codygillespie opened the pull request.

Commit:
Added validators

Added classes for validating slave and master serial data

Commit:
Name spaces

Adjsuted the namespaces of the Validator classes. Added a Protocol violation event.

corytodd commented on Apex7000_BillValidator/Serial/BaseDataValidator.cs:
```diff
@@ -0,0 +1,73 @@
+        /// Creates a new Base Data Validator
+        /// </summary>
+        /// <param name="rawData">Raw data received</param>
+        protected BaseDataValidator(byte[] rawData)
```
RawData and Data should be immutable so just a setter is fine. Then called the validation check and set the IsValid flag. That way, every call to the IsValid property doesn't cause a complete recalculation.

corytodd commented on Apex7000_BillValidator/Serial/BaseDataValidator.cs:
```diff
@@ -0,0 +1,73 @@
+            var actual = 0;
+            var expected = RawData.Last();
+
+            foreach (var b in RawData.Skip(1).Take(RawData.Length - 3))
```
Should be defensive here and make sure Length is greater than 3 so a negative index is not passed into Take.

corytodd commented on Apex7000_BillValidator/Serial/BaseDataValidator.cs:
```diff
@@ -0,0 +1,73 @@
+    /// <summary>
+    /// Base class for defining serial data validation
+    /// </summary>
+    public abstract class BaseDataValidator
```
Can these validators be internal?

codygillespie replied:
The classes that inherit BaseDataValidator can be.

BaseDataValidator was left non-internal since a BaseDataValidator is used as a property in ProtocolViolationEventArgs.

corytodd replied:
Oh okay, I didn't see that it was exposed as an event arg.

corytodd commented on Apex7000_BillValidator/Serial/MasterDataValidator.cs:
```diff
@@ -0,0 +1,130 @@
+            
+            if ((security > 0) | (orientation1 > 0) | (orientation2 > 0))
+            {
+                ValidationMessageFragments.Add("Reserved features are in use.");
```
"Reserved bites of byte 1 are not set to zero"

corytodd commented on Apex7000_BillValidator/Serial/MasterDataValidator.cs:
```diff
@@ -0,0 +1,130 @@
+
+            if (Data[2] != 0)
+            {
+                ValidationMessageFragments.Add("Reserved bit set: byte 2");
```
"Reserved byte 2 is not zero"

corytodd commented on Apex7000_BillValidator/Serial/SlaveDataValidator.cs:
```diff
@@ -0,0 +1,224 @@
+        /// </summary>
+        protected void CheckStates()
+        {
+            var slaveStates = new List<string>();
```
I would just modify States directly. Clear it at the start, .Add when bits are set.

corytodd commented on Apex7000_BillValidator/Serial/SlaveDataValidator.cs:
```diff
@@ -0,0 +1,224 @@
+        /// </summary>
+        protected void CheckEvents()
+        {
+            var slaveEvents = new List<string>();
```
I would modify Events directly. Clear at the start, .Add when bits are set.

corytodd commented on Apex7000_BillValidator/Serial/SlaveDataValidator.cs:
```diff
@@ -0,0 +1,224 @@
+        /// </summary>
+        protected void CheckErrors()
+        {
+            if ((Data[2] & 0b00000001) != 0)
```
Would be good to clear Errors at the start.

corytodd commented on Apex7000_BillValidator/Serial/SlaveDataValidator.cs:
```diff
@@ -0,0 +1,224 @@
+            {
+                ValidationMessageFragments.Add("Both state (byte 0) and error (byte 1) bytes are zero. " +
+                                               "The state byte should only equal zero if there are error bits set.");
+                return false;
```
I don't think we should return yet. Maybe just a running flag you modify. `okay = okay && SomeCheck()` so that once okay is set to false, it cannot be set to true again.

Commit:
Misc improvements

Removed Data validator constructor. Setting RawData will now automatically validate the raw data and set IsValid appropriately. Changed several validation messages. Validation will no longer hault on the first error found.

corytodd commented:
Have you had a chance to test this on real hardware yet?

codygillespie commented:
> Have you had a chance to test this on real hardware yet?

@corytodd Not yet. I am going to give it a quick test now that this PR is out of the _draft_ stage.

Commit:
Semver

1.4.2
